Add currency conversions for signing sheet - Fixes #8027

acolytec3 opened the pull request:
fixes #8027

### Summary

Displays the amount and fee associated with a transaction in the User's wallet's currency 

##### Functional
- wallet / transactions


### Steps to test
- Open Status
- Select Wallet from bottom taskbar
- Select account tile
- Select "Send"
- Fill out required transaction details
- Select "sign transaction"
- Validate that amount/fee values display a converted wallet currency amount next to them

status: ready

acolytec3 committed:
Add currency conversions for signing sheet

Signed-off-by: Andrey Shovkoplyas <motor4ik@gmail.com>

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | 9f87e03f | [#1](https://ci.status.im/job/status-react/job/prs/job/linux/job/PR-9063/1/) | 2019-09-27 20:07:44 | ~1 min | `linux` | [:page_facing_up: log](https://ci.status.im/job/status-react/job/prs/job/linux/job/PR-9063/1/consoleText) |

acolytec3 commented:
@yenda How do I get access to the Jenkins logs to see what it doesn't like?  I've run the tests and linter and both pass at this point.

cammellos commented:
@acolytec3 does not seems related to the pr, tests fails because we run out of space on jenkins (maybe just make sure they pass locally in the meantime), @jakubgs could you have a look please?

jakubgs commented:
I've cleaned up the hosts, should be good now:
```
 $ ansible ci-slave-linux -m shell -a 'df -h /'

linux-05.do-ams3.ci.misc | CHANGED | rc=0 >>
Filesystem      Size  Used Avail Use% Mounted on
/dev/vda1       310G   90G  221G  29% /

linux-04.do-ams3.ci.misc | CHANGED | rc=0 >>
Filesystem      Size  Used Avail Use% Mounted on
/dev/vda1       310G   53G  258G  17% /

linux-01.do-ams3.ci.misc | CHANGED | rc=0 >>
Filesystem      Size  Used Avail Use% Mounted on
/dev/vda1       310G   98G  213G  32% /

linux-03.do-ams3.ci.misc | CHANGED | rc=0 >>
Filesystem      Size  Used Avail Use% Mounted on
/dev/vda1       310G   71G  240G  23% /

linux-02.do-ams3.ci.misc | CHANGED | rc=0 >>
Filesystem      Size  Used Avail Use% Mounted on
/dev/vda1       310G   94G  217G  31% /
```
Please push again to rebuild everything.

acolytec3 commented:
Done, squashed and updated.

Below is a screengrab of how the currency conversion looks.

![Currency Conversion](https://user-images.githubusercontent.com/17355484/65814906-1e4a6400-e1b6-11e9-8187-2c50088b8a82.png)  

2 questions:
1) The original render in the issue shows a spacing dot of some sort between the crypto and wallet currency amounts and I wasn't sure if there was a specific character I should use as that spacer or if there is some prebuilt component to place there.  I just have a blank space currently.
2) Do you want me to insert the "~" for either or both currency fields?  Here again, the render has it displayed for the fee but not the value section.

yenda commented:
@errorists could you answer those questions?

@acolytec3 before a designer comes by to answer your question I think you should use an interpunct for now https://en.wikipedia.org/wiki/Interpunct and also be mindful of the font-color, the value is black. You can check the nested-text component and how it is used to achieve that within one text element

acolytec3 commented:
@yenda I've incorporated a nested-text element into the presentation so now the text of the wallet-currency is now black.  I think the only thing left is now I have an odd spacing issue where there's a gap between the value and the currency symbol as shown below.  Does nested-text take a padding or margin style element that would shrink this gap?  I looked at the prop functions for nested text and it looks like all the styling options are font related and not spacing.

![Currency Conv 2](https://user-images.githubusercontent.com/17355484/65824405-8476cb80-e236-11e9-957d-0bdf895ef7ae.png)

Also, since I already had the code basically written, I went ahead and added the loading spinner as well.  The bounty can still be paid out to @alexjg since I know he was planning to work on it.  Below is a screen-grab.

![Currency Conversion](https://user-images.githubusercontent.com/17355484/65824428-f8b16f00-e236-11e9-9125-956c5e998a7a.png)

errorists commented:
yes, Eric is right, both should use `~`. What I couldn't help but notice is the huuge space between value and currency, it's not right, should be a single space character. Also the space around the `•` bullet character (is that a bullet there? looks super small) should be a single space.

yenda commented on src/status_im/ui/screens/signing/views.cljs:
```diff
@@ -240,13 +245,19 @@
             :title       :t/network-fee
             :error       gas-error
-            :accessories [[acc-text fee fee-display-symbol] :chevron]
+            :accessories [[acc-text fee fee-display-symbol] "·" (if converted-fee-value [react/nested-text {:styles {:color :colors/black}} (str "~" (i18n/format-currency converted-fee-value (:code wallet-currency)))]
```
don't use `acc-text` here, if you look at the definition above in this namespace it is just a nested-text with 2 parts and here you have more. You also misused nested-text here, you should use nested-text instead of acc-text. here is an example of how to use it:

https://github.com/status-im/status-react/blob/496f3f1cc78703c3d629f9f458a0e966b04e5c61/src/status_im/ui/screens/privacy_policy/views.cljs

as you can see you give a global style and then you can have as many arguments as you want, strings or vector, the strings get the global style and the vectors are tuples of style + string, the style is merged with the global style so you only have to put what changes there

andmironov commented:
Hey! Spotted there are error tooltips on the screens, just wanted to make sure the error states stay consistent, here's an issue about those https://github.com/status-im/status-react/issues/9037#issuecomment-534986779

cc @rachelhamlin

hesterbruikman commented:
> Hey! Spotted there are error tooltips on the screens, just wanted to make sure the error states stay consistent, here's an issue about those [#9037 (comment)](https://github.com/status-im/status-react/issues/9037#issuecomment-534986779)

On the tooltip. There is a component here: https://github.com/status-im/status-react/blob/bf6b328996581c77fff437efcbf9d411e00d2ef9/src/status_im/ui/components/tooltip/styles.cljs

@yenda do you know if it's up to date, following these designs?: https://www.figma.com/file/cb4p8AxLtTF3q1L6JYDnKN15/Index?node-id=4290%3A1769

yenda commented on src/status_im/ui/screens/signing/views.cljs:
```diff
@@ -239,14 +243,26 @@
+                             " "
+                             (str (:code wallet-currency))]}
+             [separator]]
+              [list-item/list-item
```
careful with parens you slurped one too many here, list-item is not part of the accessories

acolytec3 replied:
Thanks for catching that.  I meant to change this back to WIP as I'm struggling to get the style props to cooperate for the vectors in my nested-text element where they override the global style though maybe that misplaced parens is the issue.  I'll check back in on it later tonight and see if that resolves it.

yenda commented on src/status_im/ui/screens/signing/views.cljs:
```diff
@@ -239,14 +243,26 @@
-            :error       gas-error
-            :accessories [[acc-text fee fee-display-symbol] :chevron]
+               :accessories [react/nested-text {:style {:color colors/gray}} 
+                            [{:style {:color colors/black}} (if amount (str amount) "0")]
```
CI won't accept this indent, run:
`make shell` then `lein cljfmt fix' to fix it

acolytec3 replied:
Will do!

yenda commented on src/status_im/ui/screens/signing/views.cljs:
```diff
@@ -46,16 +46,16 @@
     (multiaccounts/displayed-name contact)
     (:address contact)))
 
+(defn acc-text [txt1 txt2]
```
no point moving this

acolytec3 replied:
Sorry, human error here.  Was actually just going to suggest removing this since I was replacing the acc-text component with a more generic nested-text element in the sheets function.  From a style perspective, is it preferred to extract these kind of subcomponents out into their own function definition?  If so, I can just repurpose the existing acc-text function to incorporate the additional elements for the currency conversion.

acolytec3 replied:
I've gone ahead and removed the acc-text function definition since I'm not using it in my PR (which I think is now basically complete) and it's not used anywhere else..  Let me know if you want me to revise things further on this one.

acolytec3 commented:
@yenda @errorists I've updated the PR with the fixes for the nested-text element and seems to have resolved the spacing elements as well.  I also changed out the bullet I was using to the larger one from @errorists's comment so let me know if there are any more visual things I need to work on.

![Currency Conversion](https://user-images.githubusercontent.com/17355484/65929463-2f6ebd00-e3d0-11e9-9c5f-55775469b877.png)

errorists commented:
Looks good now, thanks @acolytec3

flexsurfer commented on src/status_im/ui/screens/signing/views.cljs:
```diff
@@ -158,7 +153,7 @@
 (defn- signing-phrase-view [phrase]
   [react/view {:align-items :center}
-   [react/text {:style {:color colors/gray :padding-bottom 8}} (i18n/label :t/signing-phrase)]
+   [react/text {:style {:color colors/gray :t-bottom 8}} (i18n/label :t/signing-phrase)]
```
what is `:t-bottom` ?

acolytec3 replied:
t for typo.  I've reverted this change.

andmironov reviewed (commented):
looks good to me as well!

churik commented:
two questions:
1) there is no value on transaction fee:
<img width="222" alt="no_value" src="https://user-images.githubusercontent.com/4557972/66220406-3eee4e80-e6cd-11e9-97b8-1fe73ad75d38.png">
Should it be added separately @flexsurfer ?
2) for some currecncies (e.g MDL) - we don't have separate symbol, so screen looks like:
<img width="185" alt="mdl_mad" src="https://user-images.githubusercontent.com/4557972/66220487-68a77580-e6cd-11e9-8ea0-dbc790460253.png">
Is it OK @andmironov ?

From my perspective PR is ready to merge.

rachelhamlin commented:
Thanks for calling these out @churik.

- Lack of symbol is okay from my POV. 
- Network fee conversion to local currency would be good for when we update that screen - [backlogged issue](https://github.com/status-im/status-react/issues/9039). Okay for now in my opinion.

Well done @acolytec3! Thanks for your contribution.
